### ISSUE TITLE
Indexing bigcount routines

### ISSUE THREAD

**Issue by wgropp** [mpi-forum/mpi-issues#334]
# Problem

The bigcount (xxx_c) C routines are not indexed (with the exception of two in discussed in the bindings chapter). These do not fit in the current indexing strategy, as the MPI API index is for the language-independent routines (for the most part).

# Proposal

Change the MPI Function index to include the XXX_c routines, and change the text describing the index to explain the _c routines, using mixed case for the C routines.  Add a way to index the _c routines (this would need to be done by indexing in mpibind, then processing these to identify the _c routines). That would be a change to mpi-user-macs.tex and buildindices.

# Changes to the Text

Only to the index.

# Impact on Implementations

None.

# Impact on Users

Ability to find the bigcount routines such as MPI_Send_c from the index.

# References

N/A

**Comment by martinruefenacht** [comment]
Is this an issue I should be resolving or is this in Latex? As far as I can tell embiggening hasn't been merged into mpi-4.x?

**Comment by wesbland** [comment]
@martinruefenacht - Check the mpi-4-rc branch. That's where everything is going until it gets voted in.

**Comment by martinruefenacht** [comment]
I see, yes Derek point out where I should be looking. I still think this is a purely Latex issue?

**Comment by wgropp** [comment]
There are several different solutions. Currently, the LaTeX macros for the bindings do *not* index the names, because there is a clear mapping to the different names, and each of the language-independent definitions has C and two Fortran bindings. The easiest option for me is to add a new LaTeX macro for the _c names, along with the mpibind command that the pythonization currently emits. A fallback is to add indexing to the mpibind macro, and then change buildindices to filter for routines that end with _c.

**Comment by martinruefenacht** [comment]
Okay, so as far as I understand, once that new macro is in place, I will need to change the generation of embiggened bindings to the new macro instead of \mpibind for C bindings. Are Fortran bindings needing to change?

**Comment by wgropp** [comment]
Actually, what would be easiest for me would be if the LaTeX emitted for the _c routines had an additional macro invocation - e.g., \mpibindbcindex{MPI\_Send\_c} . Is this easy? If not, I can add the processing to buildindices.

**Comment by martinruefenacht** [comment]
Yes, that should be easy.

Is "\mpibindbcindex{name_c}" contained inside the current \mpibind{\mpibindbcindex{...}code} or before/after(separate line)?

This whole thing goes near a hack we had to do (at last minute) with MPI_Type_get_contents and MPI_Type_get_envelope. So once a general overhaul is in order, after MPI 4.0, I will produce better/clearer code for MPI 4.1.

**Comment by martinruefenacht** [comment]
@wgropp I have done the initial work from the Python side. Let me know if it needs to change.

At the moment it outputs this:

```
\mpibind{MPI\_Send(\gb{}const~void~*buf, int~count, MPI\_Datatype~datatype, int~dest, int~tag, MPI\_Comm~comm)}
\mpibindbcindex{MPI\_Send\_c}
\mpibind{MPI\_Send\_c(\gb{}const~void~*buf, MPI\_Count~count, MPI\_Datatype~datatype, int~dest, int~tag, MPI\_Comm~comm)}
```

**Comment by wgropp** [comment]
That looks good.
For now, add

\newcommand{\mpibindbcindex}[1]{\relax}

to mpi-user-macs.tex .  I'll need to adapt buildindices together with this, but this will let the LaTeX build.

**Comment by martinruefenacht** [comment]
Done. I have created the PR to merge into mpi-4-rc which I assume is the correct target.

**Comment by wesbland** [comment]
I've been telling people to make their PRs against `mpi-4.x`, but it's fine to use the RC branch. They'll get merged back together in February anyway.

**Comment by wgropp** [comment]
@martinruefenacht I found a bug - the code is issuing \mpibindbcindex for all of the _c routines in the generated appLang-CNames.tex - this causes bogus index entries to be added.

**Comment by martinruefenacht** [comment]
@wgropp I think this is a bug in MAKE-APPLANG. Line 64:
```
re='^ *\\mpibind|^ *\\mpiemptybind'
```

I changed it to:

```
re='^ *\\mpibind{|^ *\\mpiemptybind{'
```

And also added the **{** to all the other language detections. This is a stop-gap fix. I think as a future item past MPI 4.0 we should reconsider these index generation mechanisms. They were created for handling Latex, but we can do better with directly writing an index from the Python intermediary. The motivation would be to reduce code/failure points.

**Comment by wgropp** [comment]
I agree that we need to retire the MAKE-APPLANG and other nearly 30 year old scripts, and exploit the new code. This is for the binding summary in the appendix - writing the index still needs to be done by having TeX write information about the occurrence (with the page number) and then process the lists as done in buildindices (which can certainly be improved - and changes to the LaTeX macros would also simply the process of generating the indices, especially removing the parameters from the indexed value).

**Comment by wesbland** [comment]
I believe this can be closed now
